File: packages/fundamentals/dnd/native/swift/PyreonSortable.swift
// PyreonSortable — the iOS-native port of @pyreon/dnd's `useSortable`.
//
// The web hook is pointer-driven (pragmatic-drag-and-drop over DOM events).
// That engine does NOT cross: a drag on iOS is a platform gesture, not a
// pointer-event protocol. What DOES cross is the part an author reasons about
// — the REORDER SEMANTICS: which item is lifted, which item it is over, which
// EDGE of that item the pointer is nearest, and where the moved element lands.
// That logic is pure, and `moveIndex` below is a verbatim port of the
// TypeScript `performReorder` arithmetic, so a drop at a given edge produces
// the SAME array on web, iOS and Android.
//
// The gesture layer on top is deliberately NOT a DOM re-implementation: it is
// SwiftUI's own `.draggable` / `.dropDestination`, so a drag reads as a normal
// iOS drag (lift, system preview, spring-loading). It is also the only option
// available: PMTC lowers `<For>` to a bare `ForEach`, never a `List`, so
// `ForEach.onMove` — which requires a `List` — cannot be reached from
// compiled Pyreon source.
//
// The drag PAYLOAD is the item's key as a `String`. `String` already conforms
// to `Transferable`, so a consumer's row type needs no conformance of its own.
//
// `PyreonSortableState` is pure logic plus `@Observable` — the SwiftUI view
// modifiers live below the fold, so the engine unit-tests headlessly.

import Foundation
import Observation
import SwiftUI

/// Which way the list runs. Mirrors the web `axis` option.
public enum PyreonSortAxis: String, Equatable {
    case vertical
    case horizontal
}

/// The edge of the hovered item nearest the pointer — the insert side.
/// Mirrors the web `DropEdge`.
public enum PyreonDropEdge: String, Equatable {
    case top
    case bottom
    case left
    case right
}

/// Sortable-list state: which key is lifted, which key it is over, and the
/// edge of that key. Mutating through the methods below drives a SwiftUI
/// re-render (`@Observable`).
///
/// The reactive item source is wired AFTER init via `bind` (from SwiftUI's
/// `.onAppear`), because a `@State` property initializer cannot capture the
/// view's own state — the same constraint `PyreonTableState.setData` documents.
@available(iOS 17.0, macOS 14.0, *)
@Observable
public final class PyreonSortableState<T> {
    private var itemsProvider: () -> [T] = { [] }
    private var keyOf: (T) -> String = { _ in "" }
    private var onReorderFn: ([T]) -> Void = { _ in }

    /// The key of the item being dragged, `nil` when idle.
    public private(set) var activeKey: String?
    /// The key of the item currently hovered, `nil` when none.
    public private(set) var overKey: String?
    /// The nearest edge of `overKey`, `nil` when none.
    public private(set) var currentEdge: PyreonDropEdge?

    public let axis: PyreonSortAxis

    public init(axis: PyreonSortAxis = .vertical) {
        self.axis = axis
    }

    /// Wire the reactive item source + reorder sink. Called from `.onAppear`.
    public func bind(
        items: @escaping () -> [T],
        by: @escaping (T) -> String,
        onReorder: @escaping ([T]) -> Void
    ) {
        self.itemsProvider = items
        self.keyOf = by
        self.onReorderFn = onReorder
    }

    // ── the web result surface, name-for-name ────────────────────────────────
    // `useSortable` returns `isActive` / `isOverKey` / `activeId` / `overId` /
    // `overEdge`, so the native engine spells them IDENTICALLY and returns the
    // same value shapes (edge as its raw string). That is what lets a row
    // template — `<Text opacity={s.isActive(t.id) ? 0.5 : 1}>` — compile
    // unchanged on all three targets with no compiler-side name mapping.
    public func isActive(_ key: String) -> Bool { activeKey == key }
    public func isOverKey(_ key: String) -> Bool { overKey == key }
    public func activeId() -> String? { activeKey }
    public func overId() -> String? { overKey }
    public func overEdge() -> String? { currentEdge?.rawValue }

    // ── drag lifecycle ───────────────────────────────────────────────────────
    /// The drag lifted off `key`.
    public func pickUp(_ key: String) { activeKey = key }

    /// The pointer is over `key`, nearest `edge`.
    public func dragOver(_ key: String, edge: PyreonDropEdge) {
        overKey = key
        currentEdge = edge
    }

    /// The pointer left `key`. A leave for a STALE key must not clear the live
    /// one — SwiftUI delivers the outgoing target's `isTargeted(false)` after
    /// the incoming target's `true`, so an unguarded clear blanks the
    /// highlight the user is actually over.
    public func dragLeave(_ key: String) {
        guard overKey == key else { return }
        overKey = nil
        currentEdge = nil
    }

    /// The drag ended without a drop.
    public func cancel() {
        activeKey = nil
        overKey = nil
        currentEdge = nil
    }

    /// Commit a drop of `source` onto `target` at `edge`. Returns `true` when
    /// the list actually changed, so a drop destination can report handled.
    @discardableResult
    public func drop(source: String, on target: String, edge: PyreonDropEdge) -> Bool {
        defer { cancel() }
        guard let next = reordered(dragKey: source, dropKey: target, edge: edge) else {
            return false
        }
        onReorderFn(next)
        return true
    }

    // ── the shared reorder algorithm ─────────────────────────────────────────
    /// Move the element at `from` so that it sits at `to`, where `to` is an
    /// index in the list WITHOUT the moved element. The single mutation
    /// primitive both the edge path and any displacement path route through,
    /// so every platform lands on identical arithmetic.
    public static func moveIndex(_ list: [T], from: Int, to: Int) -> [T] {
        var out = list
        let moved = out.remove(at: from)
        out.insert(moved, at: max(0, min(to, out.count)))
        return out
    }

    /// Pure reorder: move `dragKey` next to `dropKey`, inserting on `edge`'s
    /// side. Returns `nil` when the move is a no-op (same key, or either key
    /// missing) — matching the TypeScript `performReorder` guards exactly.
    ///
    /// The insert-index arithmetic is a verbatim port of the web engine; that
    /// is what makes a drop produce the same array on all three targets.
    public func reordered(dragKey: String, dropKey: String, edge: PyreonDropEdge) -> [T]? {
        if dragKey == dropKey { return nil }
        let current = itemsProvider()
        guard
            let dragIndex = current.firstIndex(where: { keyOf($0) == dragKey }),
            let dropIndex = current.firstIndex(where: { keyOf($0) == dropKey })
        else { return nil }

        let after = edge == .bottom || edge == .right
        let rawInsert: Int
        if after {
            rawInsert = dropIndex >= dragIndex ? dropIndex : dropIndex + 1
        } else {
            rawInsert = dropIndex <= dragIndex ? dropIndex : dropIndex - 1
        }
        return PyreonSortableState.moveIndex(current, from: dragIndex, to: rawInsert)
    }

    /// The edge nearest a drop `point` inside a view of `size`, on this
    /// state's axis. Vertical lists split top/bottom at the midline;
    /// horizontal lists split left/right.
    public func edgeAt(_ point: CGPoint, in size: CGSize) -> PyreonDropEdge {
        switch axis {
        case .vertical:
            return point.y < size.height / 2 ? .top : .bottom
        case .horizontal:
            return point.x < size.width / 2 ? .left : .right
        }
    }
}

// ─────────────────────────── SwiftUI binding ────────────────────────────────

/// The per-item drag/drop wiring. A `ViewModifier` (not a bare `View`
/// extension) because it needs its OWN `@State` for the measured item size:
/// the drop `location` arrives in the item's LOCAL space, and the edge is
/// meaningless without the item's height/width.
@available(iOS 17.0, macOS 14.0, *)
public struct PyreonSortableItemModifier<T>: ViewModifier {
    private let state: PyreonSortableState<T>
    private let key: String
    @State private var measured: CGSize = .zero

    public init(state: PyreonSortableState<T>, key: String) {
        self.state = state
        self.key = key
    }

    public func body(content: Content) -> some View {
        content
            .background(
                GeometryReader { proxy in
                    Color.clear
                        .onAppear { measured = proxy.size }
                        .onChange(of: proxy.size) { _, next in measured = next }
                }
            )
            // The lifted row dims — the standard iOS drag affordance.
            .opacity(state.isActive(key) ? 0.5 : 1)
            .draggable(key) {
                // The drag preview. Returning `content` would re-enter this
                // modifier; a neutral chip keeps the preview cheap.
                Text(key).padding(6)
            }
            .dropDestination(for: String.self) { payload, location in
                guard let source = payload.first else { return false }
                return state.drop(
                    source: source,
                    on: key,
                    edge: state.edgeAt(location, in: measured)
                )
            } isTargeted: { targeted in
                if targeted {
                    // The precise edge arrives with the drop location; on
                    // entry seed the leading edge so a hover highlight has
                    // something to read.
                    state.dragOver(key, edge: state.axis == .vertical ? .top : .left)
                } else {
                    state.dragLeave(key)
                }
            }
    }
}

@available(iOS 17.0, macOS 14.0, *)
extension View {
    /// Make this view a sortable item of `state`, identified by `key`.
    /// Emitted by PMTC for `ref={s.itemRef(key)}`.
    public func pyreonSortableItem<T>(
        _ state: PyreonSortableState<T>,
        key: String
    ) -> some View {
        modifier(PyreonSortableItemModifier(state: state, key: key))
    }

    /// Mark this view as the sortable CONTAINER of `state`. Emitted by PMTC
    /// for `ref={s.containerRef}`. It groups the items for assistive
    /// technology and carries the reorder hint; the per-item modifier owns the
    /// drag itself.
    public func pyreonSortableContainer<T>(
        _ state: PyreonSortableState<T>
    ) -> some View {
        accessibilityElement(children: .contain)
            .accessibilityHint(Text("Drag an item to reorder the list"))
    }
}

File: packages/fundamentals/dnd/native/kotlin/com/pyreon/runtime/PyreonSortable.kt
package com.pyreon.runtime

import androidx.compose.runtime.mutableStateOf
import kotlin.math.roundToInt

// PyreonSortable — the Android-native port of @pyreon/dnd's `useSortable`.
//
// The web hook is pointer-driven (pragmatic-drag-and-drop over DOM events).
// That engine does NOT cross: a drag on Android is a platform gesture. What
// DOES cross is the REORDER SEMANTICS — which item is lifted, which item it is
// over, which EDGE of that item is nearest, and where the moved element lands.
// `moveIndex` below is a verbatim port of the TypeScript `performReorder`
// arithmetic, and the Swift port routes through the identical primitive, so a
// reorder produces the SAME list on web, iOS and Android.
//
// This file is deliberately PURE: the only Compose dependency is
// `mutableStateOf` (so a mutation recomposes a reader), exactly like
// PyreonTableState.kt. There is no gesture code here — the long-press drag is
// EMITTED into the app's own composable by PMTC, following the same shape
// `<Press onSwipeLeft>` already uses (`pointerInput` +
// `detectDragGesturesAfterLongPress`). Keeping the gesture in the emit rather
// than the runtime is what lets `validateKotlin` type-check it against the
// mirrored Compose stubs, and lets this file be verified headlessly by
// `check-native-cosource` with no Compose-UI stub bundle at all.
//
// Android's drag model is DISPLACEMENT-based (long-press, then move by whole
// slots) where iOS is DROP-TARGET-based. Both funnel into `moveIndex`, so the
// gesture differs per platform — as it should — while the RESULT does not.

/** Which way the list runs. Mirrors the web `axis` option. */
enum class PyreonSortAxis { VERTICAL, HORIZONTAL }

/** The edge of the hovered item nearest the pointer — the insert side. */
enum class PyreonDropEdge { TOP, BOTTOM, LEFT, RIGHT }

/**
 * Sortable-list state: which key is lifted, which key it is over, and the edge
 * of that key. Backed by `mutableStateOf` so a drag recomposes the rows that
 * read `isActive` / `isOver`.
 */
class PyreonSortableState<T>(
    val axis: PyreonSortAxis = PyreonSortAxis.VERTICAL,
) {
    private var itemsProvider: () -> List<T> = { emptyList() }
    private var keyOf: (T) -> String = { "" }
    private var onReorderFn: (List<T>) -> Unit = {}

    private val _activeKey = mutableStateOf<String?>(null)
    /** The key of the item being dragged, `null` when idle. */
    val activeKey: String?
        get() = _activeKey.value

    private val _overKey = mutableStateOf<String?>(null)
    /** The key of the item currently hovered, `null` when none. */
    val overKey: String?
        get() = _overKey.value

    private val _currentEdge = mutableStateOf<PyreonDropEdge?>(null)
    /** The nearest edge of [overKey], `null` when none. */
    val currentEdge: PyreonDropEdge?
        get() = _currentEdge.value

    // Accumulated drag distance along the axis, in pixels, since pick-up, and
    // the dragged row's extent (height for a vertical list). Plain fields, not
    // Compose state: they change on every pointer sample and nothing renders
    // them directly — the derived slot delta lands in _overKey, which does.
    private var dragAccum: Float = 0f
    private var itemExtent: Float = 0f

    /** Wire the reactive item source + reorder sink. */
    fun bind(items: () -> List<T>, by: (T) -> String, onReorder: (List<T>) -> Unit) {
        itemsProvider = items
        keyOf = by
        onReorderFn = onReorder
    }

    // ── the web result surface, name-for-name ────────────────────────────────
    // `useSortable` returns `isActive` / `isOverKey` / `activeId` / `overId` /
    // `overEdge`, so the native engine spells them IDENTICALLY and returns the
    // same value shapes (edge as its lowercase string). That is what lets a row
    // template compile unchanged on all three targets with no compiler-side
    // name mapping.
    fun isActive(key: String): Boolean = activeKey == key
    fun isOverKey(key: String): Boolean = overKey == key
    fun activeId(): String? = activeKey
    fun overId(): String? = overKey
    fun overEdge(): String? = currentEdge?.name?.lowercase()

    // ── drag lifecycle ───────────────────────────────────────────────────────
    /** The long-press lifted `key`. */
    fun pickUp(key: String) {
        _activeKey.value = key
        dragAccum = 0f
        itemExtent = 0f
    }

    /**
     * Accumulate a drag sample. [delta] is this frame's movement along the
     * axis and [extent] the dragged row's size on that axis (Compose hands
     * both to the emitted gesture: the drag amount and `PointerInputScope.size`).
     *
     * A zero/negative extent is ignored rather than dividing — a row measured
     * before layout reports 0, and a divide there would send the slot delta to
     * infinity and reorder the list on the first pixel of movement.
     */
    fun dragBy(delta: Float, extent: Float) {
        if (extent > 0f) itemExtent = extent
        dragAccum += delta
        val active = activeKey ?: return
        val slots = slotDelta()
        if (slots == 0) {
            _overKey.value = null
            _currentEdge.value = null
            return
        }
        val current = itemsProvider()
        val from = current.indexOfFirst { keyOf(it) == active }
        if (from < 0) return
        val target = (from + slots).coerceIn(0, current.size - 1)
        _overKey.value = keyOf(current[target])
        _currentEdge.value = if (slots > 0) {
            if (axis == PyreonSortAxis.VERTICAL) PyreonDropEdge.BOTTOM else PyreonDropEdge.RIGHT
        } else {
            if (axis == PyreonSortAxis.VERTICAL) PyreonDropEdge.TOP else PyreonDropEdge.LEFT
        }
    }

    /** How many whole slots the accumulated drag covers. */
    private fun slotDelta(): Int =
        if (itemExtent <= 0f) 0 else (dragAccum / itemExtent).roundToInt()

    /**
     * Commit the drag. Returns `true` when the list actually changed.
     */
    fun drop(): Boolean {
        val active = activeKey
        val slots = slotDelta()
        cancel()
        if (active == null || slots == 0) return false
        val current = itemsProvider()
        val from = current.indexOfFirst { keyOf(it) == active }
        if (from < 0) return false
        val to = (from + slots).coerceIn(0, current.size - 1)
        if (to == from) return false
        onReorderFn(moveIndex(current, from, to))
        return true
    }

    /** The drag ended without a drop. */
    fun cancel() {
        _activeKey.value = null
        _overKey.value = null
        _currentEdge.value = null
        dragAccum = 0f
        itemExtent = 0f
    }

    // ── the shared reorder algorithm ─────────────────────────────────────────
    /**
     * Pure reorder: move `dragKey` next to `dropKey`, inserting on `edge`'s
     * side. Returns `null` when the move is a no-op (same key, or either key
     * missing) — matching the TypeScript `performReorder` guards exactly.
     *
     * Not on the Android gesture path (which is displacement-based) but kept
     * as the CROSS-PLATFORM contract: it is the shape the web engine and the
     * iOS drop destination both use, and the parity test asserts all three
     * agree.
     */
    fun reordered(dragKey: String, dropKey: String, edge: PyreonDropEdge): List<T>? {
        if (dragKey == dropKey) return null
        val current = itemsProvider()
        val dragIndex = current.indexOfFirst { keyOf(it) == dragKey }
        val dropIndex = current.indexOfFirst { keyOf(it) == dropKey }
        if (dragIndex < 0 || dropIndex < 0) return null

        val after = edge == PyreonDropEdge.BOTTOM || edge == PyreonDropEdge.RIGHT
        val rawInsert = if (after) {
            if (dropIndex >= dragIndex) dropIndex else dropIndex + 1
        } else {
            if (dropIndex <= dragIndex) dropIndex else dropIndex - 1
        }
        return moveIndex(current, dragIndex, rawInsert)
    }

    companion object {
        /**
         * Move the element at [from] so that it sits at [to], where [to] is an
         * index in the list WITHOUT the moved element. The single mutation
         * primitive every platform routes through.
         */
        fun <T> moveIndex(list: List<T>, from: Int, to: Int): List<T> {
            val out = list.toMutableList()
            val moved = out.removeAt(from)
            out.add(to.coerceIn(0, out.size), moved)
            return out
        }
    }
}
